cartographer/mapping/pose_extrapolator.cc

```diff
@@ -96,46 +103,47 @@ void PoseExtrapolator::UpdateVelocitiesFromPoses() {
+  while (imu_data_.size() > 1 && !timed_pose_queue_.empty() &&
          imu_data_[1].time <= timed_pose_queue_.back().time) {
     imu_data_.pop_front();
   }
 }
 
-Eigen::Quaterniond PoseExtrapolator::ExtrapolateRotation(
-    const common::Time time) {
-  common::Time current = timed_pose_queue_.back().time;
-  if (imu_data_.empty() || imu_data_.front().time >= time) {
-    return RotationFromAngularVelocity(time - current,
-                                       angular_velocity_from_poses_);
+void PoseExtrapolator::AdvanceImuTracker(const common::Time time,
+                                         ImuTracker* const imu_tracker) {
+  CHECK_GE(time, imu_tracker->time());
+  if (imu_data_.empty() || time < imu_data_.front().time) {
+    // There is no IMU data until 'time', so we advance the ImuTracker and use
+    // the angular velocities from poses and fake gravity to help 2D stability.
+    imu_tracker->Advance(time);
+    imu_tracker->AddImuLinearAccelerationObservation(Eigen::Vector3d::UnitZ());
+    imu_tracker->AddImuAngularVelocityObservation(angular_velocity_from_poses_);
+    return;
   }
-  // TODO(whess): Use the ImuTracker here?
-  // TODO(whess): Keep the last extrapolated pose.
-  Eigen::Quaterniond current_rotation;
-  auto imu_it = imu_data_.begin();
-  if (imu_it->time > current) {
-    current_rotation = RotationFromAngularVelocity(
-        imu_it->time - current, angular_velocity_from_poses_);
-    current = imu_it->time;
-  } else {
-    current_rotation = Eigen::Quaterniond::Identity();
+  if (imu_tracker->time() < imu_data_.front().time) {
+    // Advance to the beginning of 'imu_data_'.
+    imu_tracker->Advance(imu_data_.front().time);
   }
-  CHECK(imu_it != imu_data_.end());
-  CHECK(imu_it->time <= current);
-  CHECK(current < time);
-  while (current < time) {
-    common::Time next = time;
-    CHECK(imu_it != imu_data_.end());
-    auto next_it = imu_it + 1;
-    if (next_it != imu_data_.end() && next > next_it->time) {
-      next = next_it->time;
-    }
-    current_rotation *=
-        RotationFromAngularVelocity(next - current, imu_it->angular_velocity);
-    current = next;
-    imu_it = next_it;
+  auto it = std::lower_bound(
+      imu_data_.begin(), imu_data_.end(), imu_tracker->time(),
+      [](const sensor::ImuData& imu_data, const common::Time& time) {
+        return imu_data.time < time;
+      });
+  while (it != imu_data_.end() && it->time < time) {
+    imu_tracker->Advance(it->time);
+    imu_tracker->AddImuLinearAccelerationObservation(it->linear_acceleration);
+    imu_tracker->AddImuAngularVelocityObservation(it->angular_velocity);
+    ++it;
   }
-  return current_rotation;
+  imu_tracker->Advance(time);
+}
+
+Eigen::Quaterniond PoseExtrapolator::ExtrapolateRotation(
+    const common::Time time) {
+  ImuTracker imu_tracker = *imu_tracker_;
+  AdvanceImuTracker(time, &imu_tracker);
+  const Eigen::Quaterniond last_orientation = imu_tracker_->orientation();
```

Comment:
Just curious but why last_orientation is queried after AdvanceImuTracker() here.